nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -374,7 +396,16 @@ void CollisionMonitor::process(const Velocity & cmd_vel_in)
   // Fill collision_points array from different data sources
   for (std::shared_ptr<Source> source : sources_) {
     if (source->getEnabled()) {
-      source->getData(curr_time, collision_points);
+      if (!source->getData(curr_time, collision_points)) {
+        RCLCPP_WARN(get_logger(), "Invalid blocking source detected, stopping the robot");
+        Velocity stop_vel;
+        stop_vel.tw = 0.0;
+        stop_vel.x = 0.0;
+        stop_vel.y = 0.0;
+        Action robot_action{STOP, stop_vel, ""};
+        publishVelocity(robot_action);
+        return;
+      }
```

Comment:
The return will break the `process()` logic: in the tail of this routine there are some things which we can't miss.
If we just return from this, the notify action state logic won't be started, and CM won't do the change notification. So, it is better to update `CollisionMonitor::notifyActionState()` with ability to pass `action_polygon` as `nullptr` and report abnormal situation for that cases.

Additionally, we need to save action with 0-velocity to `robot_action_prev_` to have notifications to work properly; and still publish polygons, since they might be in different frames moving while robot is being stopped.

I think, here is better to set main `robot_action` local variable and go to `process()` routine's tail, calling `notifyActionState()` with null `action_polygon` that will tell this routine to report abnormal behavior.

Reply:
See last commit